Rodentia Species

portfiend opened the pull request:
## About the PR
adds Rodentia, a new species of anthropomorphic rodents (with some non-rodent-but-still-adjacent marking options). 

## Why / Balance
rodents, rabbits, bats, possums are all species that are currently not covered by our customization system but have some level of demand to them. i made sure to give this species a lot of options customization-wise to cover all of these bases. however, someone else mentioned on Discord that bats should probably also be their own species too (these rodentia "bats" are wingless) and i generally agree with that sentiment

## Technical details
Rodentia's rummage produces garbage like 70% of the time and edible food the remaining 30%. This food is either a single wedge of cheese (not a wheel), a snack, or (rarely) a single-serving meal. Keep in mind that rummaging is ~~one-time per disposal chute~~ once per disposal bin on a cooldown - not per-rodentia - so each chute on the station can only be rummaged once every 5 minutes. This combined with competition with other hungry rodentia and limited access means that rodentia are unlikely to produce lots of food. I can still adjust the numbers though

mouth storage ability grants the following:
- a 2x2 storage in their mouth, which cannot store sharp objects or animals
- being pushed, attacked for more than 2 damage, or knocked over spits out the items
- other people can tell if the rodentia has something in their mouth
- they cannot eat or speak intelligibly if something is in their mouth

the name scheme of Rodentia is "Last First", where Last is a descriptor (strong, big-eyed, trash-eating) and First is either a) a pet name (nibbles, peanut) b) a simple "first name" (monty, basil), or c) what sounds like the name of an english furry-themed town (Squeakson)

## Media

![Screenshot_149](https://github.com/user-attachments/assets/435b2aa8-a383-4332-a571-a882cd83cee7)
![Screenshot_147](https://github.com/user-attachments/assets/453018b7-96b6-4b61-964f-251ab2b9bc65)

![Screenshot_144](https://github.com/user-attachments/assets/67343917-2538-4919-b145-f34a7423adae) ![Screenshot_143](https://github.com/user-attachments/assets/9f0cac73-b4a6-4e4e-907c-1b11843633d5) ![Screenshot_142](https://github.com/user-attachments/assets/0f9d99a6-1eee-4d06-9189-920b0b71f583) ![Screenshot_141](https://github.com/user-attachments/assets/372ec668-d7cc-4974-b474-707312ed6735)

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
Rummage action on disposal chutes is now bound to the Rummager component rather than the RatKing component. If you want to give an entity the ability to rummage through trash, give them "Rummager"

RatKingRummageable objects now use a 5-minute cooldown instead of a one-time "Looted" flag. setting `RatKingRummageable.Looted = true` will still disable the object from further rummaging, but now when the object is looted the `RatKingRummageable.LastLooted` property is set instead

adds a new Naming scheme type, LastFirst, which is added to the switch statement in NamingSystem.cs

mousetraps can now be always activated with a new component, AlwaysTriggerMousetrapComponent

**Changelog**

:cl: portfiend
- add: Added Rodentia, a new species of playable rodent-like humanoids.
- tweak: Mobs that are capable of rummaging (such as the new Rodentia species) can do so every five minutes. They will find small snacks in trash bins instead of entire wheels of cheese.

## Other Notes
if you're from another server, yes you can port this

most of the discussion on this species is taking place on the DV suggestions forum in the discord but i will read replies here too

# THIS PR IS MIT-LICENSED, HAS BEEN SINCE SEPTEMBER 2024: https://github.com/DeltaV-Station/Delta-v/pull/1511#issuecomment-2339581002

portfiend committed:
Merge pull request #3 from DeltaV-Station/master

merge

portfiend committed:
add: basic rodent species

portfiend committed:
add: rodentia base sprites

portfiend committed:
add: basic rodent hand/foot markings

portfiend committed:
add: ears i got em

portfiend committed:
add: tail markings

portfiend committed:
add: big ears

by request

portfiend committed:
tweak: default markings, coloration

everything uses skin color by default when added except for these new "default" markings which use tattoo

portfiend committed:
add: the rest of the basic body part markings

portfiend committed:
add: snout textures

portfiend committed:
add: pattern markings

portfiend committed:
add: cheek detail

portfiend committed:
add: sex variations for chest marking

portfiend committed:
add: give rodentia the generic markings

portfiend committed:
add: rodentia names

portfiend committed:
add: damage, voice, mass, emotes to rodentia

portfiend committed:
add: rummager ability for rodentia

portfiend committed:
remove: temperature protection from rodentia

portfiend committed:
add: some makeup markings for rodentia

portfiend committed:
add: include rodentia in misc pools

portfiend committed:
add: rodentia guidebook entries

portfiend committed:
tweak: adjust rodentia marking points

portfiend committed:
add: rodentia pie sprite

portfiend committed:
tweak desc

portfiend committed:
tweak: change rodentia name scheme

DEATHB4DEFEAT reviewed (changes requested):
Isn't this just a Felinid with some extra markings?

portfiend commented:
> Isn't this just a Felinid with some extra markings?

yeah, i have said in multiple places that the playstyle of this is not unique enough for my tastes which is why i have been requesting suggestions here and on discord

portfiend committed:
tweak: adjust damage modifiers for rodentia

portfiend committed:
add: rodentia always trigger mouse traps

portfiend committed:
tweak: rodentia have unique speech verbs

portfiend committed:
tweak: change rodentia inventory shape

portfiend committed:
tweak: rodentia take more phys dmg overall

portfiend committed:
tweak: made rodentia hungry

portfiend committed:
add: mouth storage component on rodents

portfiend committed:
add: server mouthstoragesystem

portfiend committed:
add: force-dump mouth storage on trip

hmm the networking on this is a little ehh. but watever

portfiend committed:
add: rodentia drop mouth items on damage

dvir001 commented:
Going to the chef be like
[image](https://github.com/user-attachments/assets/bb4f8b86-c580-4717-b3cf-e54c0d2636aa)

Can probably give it an extra faction shared with the rat king and rats.

portfiend committed:
add: blocked interactions if mouth is full

portfiend committed:
fix: corrected accent oops

portfiend committed:
tweak: tense

portfiend committed:
fuck this import in particular

portfiend committed:
tweak: spit damage threshold

portfiend committed:
fix: no redundant open actions

portfiend committed:
refactor: move this action into the types file

portfiend committed:
add: rodentia get ability to hide under tables etc.

portfiend committed:
tweak: change rodentia action icons

portfiend committed:
refactor: im renaming this component and system lol

portfiend committed:
refactor: rename this event too

i was gonna refactor the whole system to be more like climbing, but the climbing system is convoluted so im just gonna do this and put it aside

portfiend committed:
tweak: fix the mouth storage icon
the number covers it

portfiend committed:
tweak: check if colliding with table before un-sneaking

portfiend committed:
fix: crawl action cooldown and popup text

portfiend committed:
add: slowdown to sneak ability

portfiend committed:
refactor: fix this namespace

portfiend committed:
tweak: rummageable objects now use a cooldown

portfiend committed:
Merge remote-tracking branch 'upstream/master' into feature/rodent-species

portfiend committed:
fix: newlines

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,146 @@
+        SubscribeLocalEvent<CrawlUnderObjectsComponent, ToggleCrawlingStateEvent>(OnAbilityToggle);
+        SubscribeLocalEvent<CrawlUnderObjectsComponent, AttemptClimbEvent>(OnAttemptClimb);
+        SubscribeLocalEvent<CrawlUnderObjectsComponent,
+            RefreshMovementSpeedModifiersEvent>(OnRefreshMovespeed);
```
```suggestion
        SubscribeLocalEvent<CrawlUnderObjectsComponent, RefreshMovementSpeedModifiersEvent>(OnRefreshMovespeed);
```

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,146 @@
+            return false;
+
+        if (TryComp<ClimbingComponent>(uid, out var climbing) && climbing.IsClimbing == true)
+            return false;
```
```suggestion
        if (component.Enabled
        	|| (TryComp<ClimbingComponent>(uid, out var climbing)
        		&& climbing.IsClimbing == true))
            return false;
```

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,146 @@
+            return false;
+
+        if (TryComp<ClimbingComponent>(uid, out var climbing) && climbing.IsClimbing == true)
+            return false;
```
```suggestion
        if (!component.Enabled
        	|| IsOnCollidingTile(uid))
        	|| (TryComp<ClimbingComponent>(uid, out var climbing)
        		&& climbing.IsClimbing == true))
            return false;
```

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,146 @@
+                if (fixtureComponent.Fixtures.TryGetValue(key, out var fixture))
+                    _physics.SetCollisionMask(uid, key, fixture, originalMask, fixtureComponent);
+            }
+        }
```
```suggestion
        if (TryComp(uid, out FixturesComponent? fixtureComponent))
            foreach (var (key, originalMask) in component.ChangedFixtures)
                if (fixtureComponent.Fixtures.TryGetValue(key, out var fixture))
                    _physics.SetCollisionMask(uid, key, fixture, originalMask, fixtureComponent);

```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/AlwaysTriggerMousetrapComponent.cs:
```diff
@@ -0,0 +1,7 @@
+[RegisterComponent]
+[NetworkedComponent]
+public sealed partial class AlwaysTriggerMousetrapComponent : Component
+{ }
```
```suggestion
public sealed partial class AlwaysTriggerMousetrapComponent : Component { }
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/CrawlUnderObjectsComponent.cs:
```diff
@@ -0,0 +1,55 @@
+}
+
+public sealed partial class ToggleCrawlingStateEvent : InstantActionEvent
+{ }
```
```suggestion
public sealed partial class ToggleCrawlingStateEvent : InstantActionEvent { }
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/RummagerComponent.cs:
```diff
@@ -0,0 +1,7 @@
+[RegisterComponent]
+[NetworkedComponent]
+public sealed partial class RummagerComponent : Component
+{ }
```
```suggestion
public sealed partial class RummagerComponent : Component { }
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/Components/MouthStorageComponent.cs:
```diff
@@ -0,0 +1,34 @@
+    public EntityUid? MouthId;
+
+    // Mimimum inflicted damage on hit to spit out items
+    [DataField("spitDamageThreshold")]
```
```suggestion
    [DataField]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/Components/MouthStorageComponent.cs:
```diff
@@ -0,0 +1,34 @@
+    public Container Mouth = default!;
+
+    [ViewVariables]
+    [DataField]
```
```suggestion
    [DataField]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/Components/MouthStorageComponent.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public const string MouthContainerId = "mouth";
+
+    [DataField("openStorageAction"), AutoNetworkedField]
```
```suggestion
    [DataField, AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/Components/MouthStorageComponent.cs:
```diff
@@ -0,0 +1,34 @@
+    [DataField, AutoNetworkedField]
+    public EntityUid? Action;
+
+    [DataField("mouthProto")]
```
```suggestion
    [DataField]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/EntitySystems/SharedMouthStorageSystem.cs:
```diff
@@ -0,0 +1,72 @@
+        if (!string.IsNullOrWhiteSpace(component.OpenStorageAction) && component.Action == null)
+        {
+            _actionsSystem.AddAction(uid, ref component.Action, component.OpenStorageAction, mouth);
+        }
```
```suggestion
        if (!string.IsNullOrWhiteSpace(component.OpenStorageAction) && component.Action == null)
            _actionsSystem.AddAction(uid, ref component.Action, component.OpenStorageAction, mouth);
```

DEATHB4DEFEAT commented on Content.Shared/Humanoid/NamingSystem.cs:
```diff
@@ -40,6 +40,9 @@ public string GetName(string species, Gender? gender = null)
                 case SpeciesNaming.FirstDashFirst:
                     return Loc.GetString("namepreset-firstdashfirst",
                         ("first1", GetFirstName(speciesProto, gender)), ("first2", GetFirstName(speciesProto, gender)));
+                case SpeciesNaming.LastFirst: // DeltaV: Rodentia name scheme
+                    return Loc.GetString("namepreset-lastfirst",
+                        ("last", GetLastName(speciesProto)), ("first", GetFirstName(speciesProto, gender)));
```
Shouldn't you just swap where you put the name lists?

portfiend replied:
first names are gendered, last names are not

DEATHB4DEFEAT commented on Resources/Locale/en-US/deltav/actions/crawl-under-objects.ftl:
```diff
@@ -0,0 +1,2 @@
+crawl-under-objects-toggle-on = Now sneaking!
+crawl-under-objects-toggle-off = Now standing!
```
```suggestion
crawl-under-objects-toggle-on = Now sneaking
crawl-under-objects-toggle-off = Now standing
```

DEATHB4DEFEAT commented on Resources/Prototypes/DeltaV/Voice/speech_emote_sounds.yml:
```diff
@@ -162,3 +162,63 @@
+    Weh:
+      collection: Weh
+    Squeak: 
+      path: /Audio/Animals/mouse_squeak.ogg
```
```suggestion
      path: /Audio/Animals/mouse_squeak.ogg

```

DEATHB4DEFEAT commented on Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml:
```diff
@@ -11,3 +11,4 @@
     Reptilian: 0.5
     SlimePerson: 0.5
     Vulpkanin: 0.5
+    Rodentia: 0.5
```
```suggestion
    Rodentia: 0.5

```

portfiend committed:
Apply suggestions from code review

Co-authored-by: DEATHB4DEFEAT <77995199+DEATHB4DEFEAT@users.noreply.github.com>
Signed-off-by: portfiend <109661617+portfiend@users.noreply.github.com>

portfiend committed:
Merge branch 'DeltaV-Station:master' into feature/rodent-species

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,141 @@
+
+    private const int HighImpassable = (int) CollisionGroup.HighImpassable;
+    private const int MidImpassable = (int) CollisionGroup.MidImpassable;
+    private const int InteractImpassable = (int) CollisionGroup.InteractImpassable;
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,141 @@
+    {
+        if (component.Enabled
+        	|| (TryComp<ClimbingComponent>(uid, out var climbing)
+        		&& climbing.IsClimbing == true))
```
```suggestion
            || (TryComp<ClimbingComponent>(uid, out var climbing)
                && climbing.IsClimbing == true))
```

DEATHB4DEFEAT commented on Content.Server/DeltaV/Abilities/CrawlUnderObjectsSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        if (!component.Enabled
+        	|| IsOnCollidingTile(uid))
+        	|| (TryComp<ClimbingComponent>(uid, out var climbing)
+        		&& climbing.IsClimbing == true))
```
```suggestion
            || IsOnCollidingTile(uid)
            || (TryComp<ClimbingComponent>(uid, out var climbing)
                && climbing.IsClimbing == true))
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/AlwaysTriggerMousetrapComponent.cs:
```diff
@@ -0,0 +1,6 @@
+namespace Content.Shared.Abilities;
+
+[RegisterComponent]
+[NetworkedComponent]
```
```suggestion
[RegisterComponent, NetworkedComponent]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/AlwaysTriggerMousetrapComponent.cs:
```diff
@@ -0,0 +1,6 @@
+
+[RegisterComponent]
+[NetworkedComponent]
+public sealed partial class AlwaysTriggerMousetrapComponent : Component { }
```
```suggestion
public sealed partial class AlwaysTriggerMousetrapComponent : Component;
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/CrawlUnderObjectsComponent.cs:
```diff
@@ -0,0 +1,54 @@
+
+[RegisterComponent]
+[NetworkedComponent]
+[AutoGenerateComponentState]
```
```suggestion
[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/CrawlUnderObjectsComponent.cs:
```diff
@@ -0,0 +1,54 @@
+    [DataField]
+    public EntityUid? ToggleHideAction;
+
+    [DataField("actionProto")]
```
```suggestion
    [DataField]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/CrawlUnderObjectsComponent.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        Enabled = enabled;
+    }
+}
```
```suggestion
[Serializable, NetSerializable]
public sealed partial class CrawlingUpdatedEvent(bool enabled = false) : EventArgs
{
    public readonly bool Enabled = enabled;
}
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Abilities/RummagerComponent.cs:
```diff
@@ -0,0 +1,6 @@
+
+[RegisterComponent]
+[NetworkedComponent]
+public sealed partial class RummagerComponent : Component { }
```
```suggestion
[RegisterComponent, NetworkedComponent]
public sealed partial class RummagerComponent : Component;
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/Components/MouthStorageComponent.cs:
```diff
@@ -0,0 +1,33 @@
+namespace Content.Shared.DeltaV.Storage.Components;
+
+[RegisterComponent, NetworkedComponent]
+[AutoGenerateComponentState]
```
```suggestion
[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/EntitySystems/SharedMouthStorageSystem.cs:
```diff
@@ -0,0 +1,70 @@
+
+    public override void Initialize()
+    {
+        base.Initialize();
```
```suggestion
        base.Initialize();

```

DEATHB4DEFEAT commented on Content.Shared/DeltaV/Storage/EntitySystems/SharedMouthStorageSystem.cs:
```diff
@@ -0,0 +1,70 @@
+        if (args.DamageDelta == null || !args.DamageIncreased)
+            return;
+
+        if (args.DamageDelta.GetTotal() < component.SpitDamageThreshold)
```
```suggestion
        if (args.DamageDelta == null
            || !args.DamageIncreased
            || args.DamageDelta.GetTotal() < component.SpitDamageThreshold)
```

portfiend committed:
Apply suggestions from code review

Co-authored-by: DEATHB4DEFEAT <77995199+DEATHB4DEFEAT@users.noreply.github.com>
Signed-off-by: portfiend <109661617+portfiend@users.noreply.github.com>

portfiend committed:
fix: collision group variavbles

portfiend committed:
Merge branch 'DeltaV-Station:master' into feature/rodent-species

portfiend committed:
add: proper "full body" sprite

portfiend committed:
add: move blaze to head folder, "round face" marking

portfiend committed:
Merge branch 'feature/rodent-species' of https://github.com/portfiend/Delta-v into feature/rodent-species

portfiend committed:
add: whisker markings

DEATHB4DEFEAT commented on Resources/Locale/en-US/deltav/markings/rodentia.ftl:
I think single `#`s are technically invalid Fluent, but RT doesn't care

DEATHB4DEFEAT reviewed (approved):
I'll let Deltanedas or someone decide on if they want this/features of this or whatever, but the code looks fine.

portfiend committed:
tweal: rodentia are 50% -> 33% hungrier

1.5x seemed excessive

portfiend commented:
ready for merge when maintainers approve it

deltanedas commented on Content.Server/DeltaV/Storage/EntitySystems/MouthStorageSystem.cs:
```diff
@@ -0,0 +1,55 @@
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var subject = Identity.Entity(uid, entityManager);
+            args.PushMarkup(Loc.GetString("mouth-storage-examine-condition-occupied", ("entity", subject)));
+        }
```
1. use EntitySystem's EntityManager field instead of resolving it
2. move this to shared

deltanedas commented on Content.Server/Mousetrap/MousetrapSystem.cs:
```diff
@@ -39,6 +40,10 @@ private void OnUseInHand(EntityUid uid, MousetrapComponent component, UseInHandE
     private void OnStepTriggerAttempt(EntityUid uid, MousetrapComponent component, ref StepTriggerAttemptEvent args)
     {
+        // DeltaV: Entities with this component always trigger mouse traps, even if wearing shoes
+        if (TryComp(args.Tripper, out AlwaysTriggerMousetrapComponent? _) == true)
```
use HasComp

deltanedas commented on Content.Shared/Storage/Components/DumpableComponent.cs:
```diff
@@ -11,6 +11,20 @@ public sealed partial class DumpableDoAfterEvent : SimpleDoAfterEvent
 }
 
+// DeltaV: Event that dumps the contents of a container with no wait
+public sealed partial class DumpContentsEvent : EntityEventArgs
```
remove this event and just have the method be public api on dumpable system

deltanedas commented on Content.Shared/Storage/EntitySystems/DumpableSystem.cs:
```diff
@@ -139,27 +140,44 @@ private void StartDoAfter(EntityUid storageUid, EntityUid targetUid, EntityUid u
+    }
+
+    // DeltaV: Refactor to allow dumping that doesn't require a verb
+    private void DumpContents(EntityUid uid, DumpableComponent component, EntityUid? target, EntityUid user)
```
make it public and have component be optional and resolved
instead of raising meme event `_dumpable.DumpContents(uid, null, null, user);`

deltanedas commented on Resources/Prototypes/DeltaV/Body/Parts/rodentia.yml:
```diff
@@ -0,0 +1,186 @@
+  parent: PartRodentia
+  components:
+  - type: Sprite
+    netsync: false
```
remove all netsync falses, put the sprite+icon paths on PartRodentia

deltanedas commented on Resources/Prototypes/DeltaV/Entities/Mobs/Player/rodentia.yml:
```diff
@@ -0,0 +1,34 @@
+- type: entity
```
this and BaseMobRodentia yml have huge duplication with BaseSpeciesOrganic, remove everything thats inherited

portfiend committed:
refactor(mouthStorage): move OnExamined from server to shared

portfiend committed:
refactor(mousetrap): use HasComp

portfiend committed:
refactor(mouthStorage): use a public api instead of event here

portfiend committed:
refactor(yml): KILL REPETITION

portfiend committed:
refactor(parts/rodentia): KILL REPETITION

deltanedas reviewed (approved):
code looks good now but needs lyndos approval too

Lyndomen commented:
ill get back to you in 3 business days

Verbalase commented:
> ill get back to you in 3 business days

what did bro really mean by this

Lyndomen reviewed (approved):
game director  approved

ghost commented:
> > ill get back to you in 3 business days
> 
> what did bro really mean by this

watch this you goober!!

ghost commented:
THE MICE ARE COMING!!

jimmy12or commented:
Yay I can make my bun!

portfiend commented:
hi also when i wrote "forks are free to port and use this" i really did mean it, which is to say if merely me giving the word is enough to will it into reality then the additions and code enclosed in this PR are licensed under the MIT license thanks
